include/ginkgo/core/matrix/csr.hpp

```diff
@@ -249,28 +268,46 @@ class Csr : public EnableLinOp<Csr<ValueType, IndexType>>,
             // if the number of stored elements is larger than 1e6 or
             // the maximum number of stored elements per row is larger than
             // 64, use load_balance otherwise use classical
+            auto host_mtx_exec = mtx_row_ptrs.get_executor()->get_master();
+            bool is_mtx_on_host{host_mtx_exec == mtx_row_ptrs.get_executor()};
+            Array<index_type> row_ptrs_host(host_mtx_exec);
+            const index_type *row_ptrs{};
+            if (is_mtx_on_host) {
+                row_ptrs = mtx_row_ptrs.get_const_data();
+            } else {
+                row_ptrs_host = mtx_row_ptrs;
+                row_ptrs = row_ptrs_host.get_const_data();
+            }
             const auto num_rows = mtx_row_ptrs.get_num_elems() - 1;
-            Array<index_type> host_row_ptrs(
-                mtx_row_ptrs.get_executor()->get_master());
-            host_row_ptrs = mtx_row_ptrs;
-            const auto row_val = host_row_ptrs.get_const_data();
-            if (row_val[num_rows] > static_cast<index_type>(1e6)) {
-                std::make_shared<load_balance>(nwarps_)->process(host_row_ptrs,
-                                                                 mtx_srow);
-                this->set_name("load_balance");
+            if (row_ptrs[num_rows] > static_cast<index_type>(1e6)) {
```

Comment:
Is 1e6 a suitable number for both AMD and CUDA ?

Reply:
I have no idea, but it also does not really matter since this PR is just aiming to improve the current strategies. #373 will (probably) address that.

Reply:
Yes, that should be the work of #373.

Reply:
Ah, okay that makes sense.